fix(nuxt): ensure all dir parts are present in component name

baiwusanyu-c opened the pull request:
<!---
☝️ PR title should follow conventional commits (https://conventionalcommits.org)

Please carefully read the contribution docs before creating a pull request
 👉 https://nuxt.com/docs/community/contribution
-->

### 🔗 Linked issue

<!-- Please ensure there is an open issue and mention its number as #123 -->

### ❓ Type of change

<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->

- [ ] 📖 Documentation (updates to the documentation, readme or JSdoc annotations)
- [x] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description
close: #20613 
<!-- Describe your changes in detail -->
<!-- Why is this change required? What problem does it solve? -->
<!-- If it resolves an open issue, please link to the issue here. For example "Resolves #1337" -->

### 📝 Checklist

<!-- Put an `x` in all the boxes that apply. -->
<!-- If your change requires a documentation PR, please link it appropriately -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

baiwusanyu-c committed:
fix(nuxt): fix scan components error

baiwusanyu-c committed:
Merge branch 'main' into bwsy/fix/scan-components

baiwusanyu-c commented:
close: #20613

danielroe committed:
refactor: extract `resolveComponentName` util

danielroe committed:
test: add some tests

danielroe reviewed (commented):
This looks great - thank you! ❤️ 

I've extracted your code into a separate helper function and wrote some unit tests to describe the behaviour I _think_ we are expecting, and which isn't entirely working yet... What do you think?

baiwusanyu-c committed:
Merge branch 'main' into bwsy/fix/scan-components

baiwusanyu-c committed:
fix(nuxt): updated code

baiwusanyu-c committed:
Merge branch 'main' into bwsy/fix/scan-components

baiwusanyu-c committed:
fix(nuxt): remove unuseful test code

baiwusanyu-c committed:
Merge remote-tracking branch 'origin/bwsy/fix/scan-components' into bwsy/fix/scan-components

baiwusanyu-c commented:
@danielroe Thank you very much for your help. I think I misunderstood his rules and logic. Now I think we need to find the judgment condition and complete the prefixPart.

danielroe commented on packages/nuxt/test/scan-components.test.ts:
```diff
@@ -230,3 +243,21 @@ it('components:scanComponents', async () => {
+const tests: Array<[string, string[], string]> = [
+  ['WithClientOnlySetup', ['Client'], 'ClientWithClientOnlySetup'],
+  ['Item', ['Item', 'Holder', 'Item'], 'ItemHolderItem'],
+  ['ItemHolder', ['Item', 'Holder', 'Item'], 'ItemHolderItem'],
```
Oh, I think I might have added a wrong test  case here... 🤔

baiwusanyu-c replied:
> Oh, I think I might have added a wrong test case here... 🤔

oh? 🤣, I was really a little confused when I modified it, I thought there were rules I didn't know. So what are the rules now?
These three are the output results of the use cases that failed to run the test before, I think only the second one is wrong
```
1.['ItemHolder', ['Item', 'Holder', 'Item'], 'ItemHolderItemHolder'],
2.['ItemHolderItem', ['Item', 'Holder', 'Item'], 'ItemHolderItemHolderItem'], // should be ItemHolderItem
3.['Item', ['Item', 'Item'], 'ItemItem'],
```

danielroe replied:
```suggestion
  ['ItemHolder', ['Item', 'Holder', 'Item'], 'ItemHolderItemHolder'],
```

danielroe commented on packages/nuxt/test/scan-components.test.ts:
```diff
@@ -230,3 +243,21 @@ it('components:scanComponents', async () => {
+const tests: Array<[string, string[], string]> = [
+  ['WithClientOnlySetup', ['Client'], 'ClientWithClientOnlySetup'],
+  ['Item', ['Item', 'Holder', 'Item'], 'ItemHolderItem'],
+  ['ItemHolder', ['Item', 'Holder', 'Item'], 'ItemHolderItem'],
```
```suggestion
  ['ItemHolder', ['Item', 'Holder', 'Item'], 'ItemHolderItemHolderItem'],
```

danielroe committed:
test: update test assertion

baiwusanyu-c committed:
fix(nuxt): updated code

baiwusanyu-c committed:
Merge branch 'main' into bwsy/fix/scan-components

baiwusanyu-c committed:
fix(nuxt): updated code

baiwusanyu-c committed:
Merge remote-tracking branch 'origin/bwsy/fix/scan-components' into bwsy/fix/scan-components

baiwusanyu-c committed:
Merge branch 'main' into bwsy/fix/scan-components

baiwusanyu-c committed:
Merge branch 'main' into bwsy/fix/scan-components

danielroe commented on packages/nuxt/src/components/scan.ts:
```diff
@@ -137,3 +121,38 @@ export async function scanComponents (dirs: ComponentsDir[], srcDir: string): Pr
+   */
+  const fileNameParts = splitByCase(fileName)
+
+  // e.g Item/Holder/Item/ItemHolderItem.vue -> ItemHolderItem
+  if (fileNameParts.join('').toLowerCase() === prefixParts.join('').toLowerCase()) {
+    return pascalCase(fileNameParts)
+  }
+
+  // e.g Item/Item/Item.vue -> Item
+  if ([...new Set(prefixParts)].length === 1 &&
+    (fileNameParts[0] || '').toLowerCase() === (prefixParts[0] || '').toLowerCase()) {
+    return pascalCase(fileNameParts)
+  }
```
I'm worried these two sections are too 'specific' to the test cases I provided rather than reproducible as an algorithm.

I think:
* let's revert back to your original proposed change without this kind of 'special-case' handling
* update the test cases to accurately describe the previous behaviour with the exception of the bug reported initially (so we don't end up _breaking_ anyone's expected component resolution)

baiwusanyu-c committed:
Merge branch 'main' into bwsy/fix/scan-components

baiwusanyu-c committed:
fix(nuxt): updated code

baiwusanyu-c commented:
I added some tests based on the bug(` ['Item', ['Item', 'Holder', 'Item'], 'ItemHolderItem']`) and the test case you provided, and ran the original code and the original commit (3d6af90) separately, and found that they didn't pass the tests. 
| fileName | prefixParts     | expected | original code | 3d6af90|
|-----|----------|---------|---------|---------|
| WithClientOnlySetup | ['Client'] | ClientWithClientOnlySetup | ClientWithClientOnlySetup✅ | ClientWithClientOnlySetup ✅ |
| ItemHolderItem | ['Item', 'Holder', 'Item'] | ItemHolderItem| ItemHolderItem✅ | ItemHolderItemHolderItem❌ |
| Item | ['Item'] | Item| Item✅ | Item ✅ |
| Item | ['Item','Item'] | Item| Item✅ | ItemItem❌ |
| ItemTest | ['Item', 'Test'] | ItemTest | ItemTest✅ | ItemTest ✅ |
| ThingItemTest| ['Item', 'Thing'] | ThingItemTest| ThingItemTest✅ | ThingItemTest✅ |
| Item | ['Thing', 'Item']| ThingItem| ThingItem✅ | ThingItem✅ |
| Item| ['Item', 'Holder', 'Item'] | ItemHolderItem| Item❌ | ItemHolderItem✅ |
| ItemHolder| ['Item', 'Holder', 'Item'] | ItemHolderItemHolder | ItemHolder❌ | ItemHolderItemHolder✅ |
| ThingItemTest| ['Item', 'Thing', 'Foo'] | ItemThingFooThingItemTest| ItemThingItemTest❌ | ItemThingItemTest❌ |
| ItemIn| ['Item', 'Holder', 'Item', 'In'] | ItemHolderItemIn | ItemIn✅ | ItemHolderItemIn✅ |
| Item| ['Item', 'Holder', 'Test'] | ItemHolderTestItem| Item❌ | Item❌ |
| ItemHolderItem| ['Item', 'Holder', 'Item', 'Holder']| ItemHolderItemHolderItem| ItemHolderItem❌ | ItemHolderItemHolderItem✅ |

Especially `['Item', ['Item', 'Holder', 'Test'], 'ItemHolderTestItem']` and `['ThingItemTest', ['Item', 'Thing', 'Foo'], 'ItemThingFooThingItemTest ']`These two test cases.

So I used `startsWith` to modify the matching implementation, but he still failed the test of `['Item', ['Item', 'Item'], 'Item']`, so I added some judgments to ensure It passes the test and behaves the same as the original code
